### PR TITLE
343 issue refactor schema registry sampler

### PR DESCRIPTION
**Pull request by nuriaricofr**

### CODE REVIEW

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ParsedSchemaAdapter.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.sngular.kloadgen.sampler.schemaregistry.adapter.impl;
+
+public abstract class ParsedSchemaAdapter {
+  public  abstract <T> T getType();
```
[the remark this answers is not in the document]
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ParsedSchemaAdapter.java:4]: changed

File: src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java [hunk 2]
```diff
@@ -166,8 +169,9 @@ public static Descriptors.Descriptor buildProtoDescriptor(final ProtoFileElement
             schemaBuilder.addSchema(importedSchema);
           }
         } else {
-          final var importedProtobufSchema = (ProtobufSchema) JMeterHelper.getParsedSchema(getSubjectName(importedClass, metadata),
-                                                                                           JMeterContextService.getContext().getProperties());
+          final ParsedSchemaAdapter protoFileElement = getParsedSchema(getSubjectName(importedClass, metadata),
```
[the remark this answers is not in the document]
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java:172]: done

File: src/main/java/com/sngular/kloadgen/processor/objectcreatorfactory/impl/ProtobufObjectCreatorFactory.java [hunk 3]
```diff
@@ -46,7 +51,12 @@ public ProtobufObjectCreatorFactory(final Object schema, final BaseSchemaMetadat
       this.schema = SchemaProcessorUtils.buildProtoDescriptor((ProtoFileElement) ((ParsedSchema) schema).rawSchema(), metadata);
     } else if (schema instanceof ProtoFileElement) {
       this.schema = SchemaProcessorUtils.buildProtoDescriptor((ProtoFileElement) schema, metadata);
-    } else {
+    } else if (schema instanceof BaseParsedSchema) {
+      BaseParsedSchema schemaParse = (BaseParsedSchema) schema;
+      ParsedSchemaAdapter adapterParse = schemaParse.getParsedSchemaAdapter();
+      final Object schemaParsed = adapterParse.getRawSchema();
+      this.schema = SchemaProcessorUtils.buildProtoDescriptor((ProtoFileElement) schemaParsed, metadata);
+  } else {
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/processor/objectcreatorfactory/impl/ProtobufObjectCreatorFactory.java:59]: Final variables and unindented else, same as in avro

File: src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java [hunk 4]
```diff
@@ -6,6 +6,8 @@
 
 package com.sngular.kloadgen.processor.util;
 
+import static com.sngular.kloadgen.util.JMeterHelper.getParsedSchema;
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java:9]: This static import can be avoided

File: src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java [hunk 5]
```diff
@@ -150,6 +152,7 @@ public static String[] splitAndNormalizeFullFieldName(final String fullFieldName
     return Arrays.stream(fields).map(field -> field.replaceAll("\\[.*]", "")).toArray(String[]::new);
   }
 
+  @SuppressWarnings("checkstyle:SingleSpaceSeparator")
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java:155]: Why is this suppressing a checkstyle warn?

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package com.sngular.kloadgen.sampler.schemaregistry.adapter.impl;
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java:1]: New files are missing the mozilla licence notice

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+public class ApicurioParsedSchemaMetadata extends ParsedSchemaAdapter {
+
+  private final String type;
+
+  private final Object schema;
+
+  private final String rawSchema;
+
+  public ApicurioParsedSchemaMetadata(final ApicurioParsedSchema apicurioParsedSchema) {
+    this.type = apicurioParsedSchema.getType().toString();
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java:16]: Type is already a string

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Override
+  public String getType() {
+    return this.type;
+  }
+
+  @Override
+  public Object getRawSchema() {
+    return this.rawSchema;
+  }
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java:29]: These overrides don't match the generics of the superclass, is it intentional?

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/AvroSchemaAdapter.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  private List<SchemaReference> references;
+  private Map<String, String> resolvedReferences;
+  private boolean isNew;
+
+  private transient int hashCode = NO_HASHCODE;
+
+  private static final int NO_HASHCODE = Integer.MIN_VALUE;
+
+ private AvroSchemaAdapter(AvroSchema avroSchema){
+   super();
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/AvroSchemaAdapter.java:26]: Does this super call without arguments do anything?

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/AvroSchemaAdapter.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+   this.canonicalString = avroSchema.canonicalString();
+   this.schemaObj =  avroSchema.rawSchema();
+   this.version= avroSchema.version();
+   this.references= avroSchema.references();
+   this.resolvedReferences=avroSchema.resolvedReferences();
+   this.isNew= avroSchema.isNew();
+ }
+
+ public static AvroSchemaAdapter parse(AvroSchema avroSchema){
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/AvroSchemaAdapter.java:35]: This parameter should be final

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/AvroSchemaAdapter.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  public static final String TYPE = "AVRO";
+
+  private Schema schemaObj;
+  private String canonicalString;
+  private Integer version;
+  private List<SchemaReference> references;
+  private Map<String, String> resolvedReferences;
+  private boolean isNew;
+
+  private transient int hashCode = NO_HASHCODE;
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/AvroSchemaAdapter.java:21]: Several of these are unused. You should also run this file through the formatter

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/BaseParsedSchema.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package com.sngular.kloadgen.sampler.schemaregistry.adapter.impl;
+
+import org.apache.avro.Schema.Parser;
+
+public class BaseParsedSchema <T extends ParsedSchemaAdapter> extends Parser {
+
+  private final T parsedSchemaAdapter;
+
+  public BaseParsedSchema(T parsedSchemaAdapter) {this.parsedSchemaAdapter = parsedSchemaAdapter;}
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/BaseParsedSchema.java:9]: Parameter should be final, and we're missing some newlines

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ConfluentParsedSchemaMetadata.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
+package com.sngular.kloadgen.sampler.schemaregistry.adapter.impl;
+
+import java.util.List;
+
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.kafka.schemaregistry.client.rest.entities.SchemaReference;
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ConfluentParsedSchemaMetadata.java:6]: Two unused imports

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ConfluentParsedSchemaMetadata.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Override
+  public String getType() {
+    return this.schemaType;
+  }
+
+  @Override
+  public Object getRawSchema() {
+    return this.rawSchema;
+  }
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ConfluentParsedSchemaMetadata.java:64]: Same thing with the signature of the override not matching generics

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ConfluentParsedSchemaMetadata.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+  private String canonicalString;
+
+  private  Object rawSchema;
+
+  private ConfluentParsedSchemaMetadata(ParsedSchema parsedSchema){
+    this.schemaType = parsedSchema.schemaType();
+    this.name = parsedSchema.name();
+    this.canonicalString = parsedSchema.canonicalString();
+    this.rawSchema = parsedSchema.rawSchema();
+  }
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ConfluentParsedSchemaMetadata.java:27]: Some of these methods should have final parameters and go through the formatter

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ParsedSchemaAdapter.java [hunk 16]
```diff
@@ -0,0 +1,7 @@
+package com.sngular.kloadgen.sampler.schemaregistry.adapter.impl;
+
+public abstract class ParsedSchemaAdapter {
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ParsedSchemaAdapter.java:3]: This abstract class only defines function signatures, should it be an interface instead?

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/impl/ConfluentSchemaRegistry.java [hunk 17]
```diff
@@ -75,23 +79,26 @@ public BaseSchemaMetadata<ConfluentSchemaMetadata> getLatestSchemaMetadata(Strin
     }
   }
 
-  public Object getSchemaBySubject(String subjectName) {
+  public BaseParsedSchema<ConfluentParsedSchemaMetadata> getSchemaBySubject(String subjectName) {
     try {
       final ConfluentSchemaMetadata schemaMetadata = ConfluentSchemaMetadata.parse(this.schemaRegistryClient.getLatestSchemaMetadata(subjectName));
-      return this.schemaRegistryClient.getSchemaBySubjectAndId(subjectName, schemaMetadata.getId());
+      ParsedSchema parsedSchema= this.schemaRegistryClient.getSchemaBySubjectAndId(subjectName, schemaMetadata.getId());
+      ParsedSchemaAdapter parsedSchemaAdapter = ConfluentParsedSchemaMetadata.parse(parsedSchema);
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/impl/ConfluentSchemaRegistry.java:86]: Formatting and final variables

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java [hunk 18]
```diff
@@ -1,17 +1,31 @@
 package com.sngular.kloadgen.sampler.schemaregistry.schema;
 
+import com.sngular.kloadgen.sampler.schemaregistry.adapter.impl.ApicurioParsedSchemaMetadata;
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:3]: Unused import

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java [hunk 19]
```diff
@@ -1,17 +1,31 @@
...
-public class ApicurioParsedSchema {
+public class ApicurioParsedSchema extends ParsedSchemaAdapter {
 
   private String type;
 
   private Object schema;
 
   private String rawSchema;
+
+  @SuppressWarnings("UnnecessarilyFullyQualified")
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:20]: Another strange warn suppression

File: src/main/java/com/sngular/kloadgen/util/JMeterHelper.java [hunk 20]
```diff
@@ -4,8 +4,10 @@
 import java.util.Map;
 import java.util.Properties;
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/util/JMeterHelper.java:5]: Not from this PR, but this file is missing a license notice

File: src/main/java/com/sngular/kloadgen/util/JMeterHelper.java [hunk 21]
```diff
@@ -16,30 +18,33 @@ public final class JMeterHelper {
   private JMeterHelper() {
   }
 
-  public static Object getParsedSchema(final String subjectName, final Properties properties) {
+  public static BaseParsedSchema getParsedSchema(final String subjectName, final Properties properties) {
     final Map<String, String> originals = new HashMap<>();
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/util/JMeterHelper.java:23]: Are the contents of this map ever used?

**Review by 5uso-sng — COMMENTED** [review]
Mostly just style issues, putting the project through the formatter and making checkstyle fixes should deal with most of it

File: src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java [hunk 22]
```diff
@@ -68,11 +70,10 @@ public final Pair<String, List<FieldValueMapping>> flatPropertiesList(final Stri
     final List<FieldValueMapping> attributeList = new ArrayList<>();
     final HashMap<String, TypeElement> nestedTypes = new HashMap<>();
     String registryName = properties.getProperty(SchemaRegistryKeyHelper.SCHEMA_REGISTRY_NAME);
+    ParsedSchemaAdapter parsedSchemaAdapter = schemaParsed.getParsedSchemaAdapter();
+    schemaType = parsedSchemaAdapter.getType();
     if (Objects.nonNull(registryName) && registryName.equalsIgnoreCase(SchemaRegistryEnum.APICURIO.toString())) {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java:75]: Conditional can be simplified as SchemaRegistryEnum.APICURIO.toString().equalsIgnoreCase(registryName)

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+@Getter
+public class ApicurioParsedSchemaMetadata extends ParsedSchemaAdapter {
+
+  private final String type;
+
+  private final Object schema;
+
+  private final String rawSchema;
+
+  public ApicurioParsedSchemaMetadata(final ApicurioParsedSchema apicurioParsedSchema) {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java:15]: Constructor argument must be an io.apicurio.registry.resolver.ParsedSchema object type

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java [hunk 24]
```diff
@@ -1,17 +1,31 @@
...
+  private ApicurioParsedSchema(final ApicurioParsedSchema apicurioParsedSchema){
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:21]: Why does the constructor need an ApicurioParsedSchema argument to instantiate an object of the same type?

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java [hunk 25]
```diff
@@ -1,17 +1,31 @@
...
+import com.sngular.kloadgen.sampler.schemaregistry.adapter.impl.ParsedSchemaAdapter;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 
 @Getter
 @Setter
 @NoArgsConstructor
-public class ApicurioParsedSchema {
+public class ApicurioParsedSchema extends ParsedSchemaAdapter {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:12]: I think that this class is not necessary. It is only used as argument to instantiate ApicurioParsedSchemaMetadata object, but that argumento must be replace by io.apicurio.registry.resolver.ParsedSchema object type

**Review by rabad-sng — COMMENTED** [review]
ApicurioParsedSchema is an unnecesarry class. Please, remove it and replace its use with the io.apicurio.registry.resolver.ParsedSchema object type.

File: src/main/java/com/sngular/kloadgen/extractor/extractors/AvroExtractor.java [hunk 26]
```diff
@@ -8,6 +8,8 @@
 
 import com.sngular.kloadgen.model.FieldValueMapping;
 import com.sngular.kloadgen.randomtool.random.RandomObject;
+import com.sngular.kloadgen.sampler.schemaregistry.adapter.impl.BaseParsedSchema;
+import com.sngular.kloadgen.sampler.schemaregistry.adapter.impl.ParsedSchemaAdapter;
```
[the remark this answers is not in the document]
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/extractor/extractors/AvroExtractor.java:12]: changed

File: src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java [hunk 27]
```diff
@@ -68,11 +70,10 @@ public final Pair<String, List<FieldValueMapping>> flatPropertiesList(final Stri
...
-      ApicurioParsedSchema apicurioParsedSchema = (ApicurioParsedSchema) schemaParsed;
-      Object schema = apicurioParsedSchema.getSchema();
-
-      schemaType = apicurioParsedSchema.getType();
+      Object schema = ((ApicurioParsedSchema) parsedSchemaAdapter).getSchema();
```
[the remark this answers is not in the document]
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java:76]: changed

File: src/main/java/com/sngular/kloadgen/processor/objectcreatorfactory/impl/AvroObjectCreatorFactory.java [hunk 28]
```diff
@@ -45,6 +48,15 @@ public AvroObjectCreatorFactory(final Object schema, final BaseSchemaMetadata<?
       this.schema = (Schema) ((ParsedSchema) schema).rawSchema();
     } else if (schema instanceof Schema) {
       this.schema = (Schema) schema;
+    } else if (schema instanceof BaseParsedSchema) {
+      BaseParsedSchema schemaParse = (BaseParsedSchema) schema;
+      ParsedSchemaAdapter adapterParse = schemaParse.getParsedSchemaAdapter();
+      if (adapterParse instanceof ApicurioParsedSchema) {
+        this.schema = (Schema) ((ApicurioParsedSchema) adapterParse).getSchema();
+      }
+      else{
```
[the remark this answers is not in the document]
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/processor/objectcreatorfactory/impl/AvroObjectCreatorFactory.java:57]: changed

File: src/main/java/com/sngular/kloadgen/processor/objectcreatorfactory/impl/ProtobufObjectCreatorFactory.java (hunk 3 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/processor/objectcreatorfactory/impl/ProtobufObjectCreatorFactory.java:59]: changed

File: src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java (hunk 4 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/processor/util/SchemaProcessorUtils.java:9]: changed

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/impl/ConfluentSchemaRegistry.java (hunk 17 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/impl/ConfluentSchemaRegistry.java:86]: changed

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java (hunk 18 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:3]: changed

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java (hunk 19 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:20]: deleted

File: src/main/java/com/sngular/kloadgen/util/JMeterHelper.java (hunk 21 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/util/JMeterHelper.java:23]: DELETED

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java (hunk 23 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/adapter/impl/ApicurioParsedSchemaMetadata.java:15]: done

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java (hunk 24 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:21]: changed

File: src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java (hunk 25 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/sampler/schemaregistry/schema/ApicurioParsedSchema.java:12]: class deleted

File: src/main/java/com/sngular/kloadgen/util/JMeterHelper.java (hunk 20 above)
Comment (nuriaricofr) [reply src/main/java/com/sngular/kloadgen/util/JMeterHelper.java]: done

**Review by jemacineiras — APPROVED** [review]

**Review by 5uso-sng — COMMENTED** [review]
`BaseParsedSchema`, `ConfluentParsedSchema`, and `ParsedSchemaAdapter` are still missing license notices. Other than that, seems good

**Review by rabad-sng — APPROVED** [review]
Proposed changes reviewed

**Pull request merged**